modules/arrow/src/geoarrow/convert-geoarrow-to-binary-geometry.ts

```diff
@@ -264,45 +265,53 @@ function getBinaryGeometriesFromChunk(
  * @param primitivePolygonIndices Indices within positions of the start of each primitive Polygon/ring
  * @param flatCoordinateArray Array of x, y or x, y, z positions
  * @param nDim - number of dimensions per position
- * @returns
+ * @returns triangle indices or null if invalid polygon and earcut fails
  */
 export function getTriangleIndices(
   polygonIndices: Uint16Array,
   primitivePolygonIndices: Int32Array,
   flatCoordinateArray: Float64Array,
   nDim: number
-): Uint32Array {
-  let primitiveIndex = 0;
-  const triangles: number[] = [];
-  // loop polygonIndices to get triangles
-  for (let i = 0; i < polygonIndices.length - 1; i++) {
-    const startIdx = polygonIndices[i];
-    const endIdx = polygonIndices[i + 1];
-    // get subarray of flatCoordinateArray
-    const slicedFlatCoords = flatCoordinateArray.subarray(startIdx * nDim, endIdx * nDim);
-    // get holeIndices for earcut
-    const holeIndices: number[] = [];
-    while (primitivePolygonIndices[primitiveIndex] < endIdx) {
-      if (primitivePolygonIndices[primitiveIndex] > startIdx) {
-        holeIndices.push(primitivePolygonIndices[primitiveIndex] - startIdx);
+): Uint32Array | null {
+  try {
+    let primitiveIndex = 0;
+    const triangles: number[] = [];
+    // loop polygonIndices to get triangles
+    for (let i = 0; i < polygonIndices.length - 1; i++) {
+      const startIdx = polygonIndices[i];
+      const endIdx = polygonIndices[i + 1];
+      // get subarray of flatCoordinateArray
+      const slicedFlatCoords = flatCoordinateArray.subarray(startIdx * nDim, endIdx * nDim);
+      // get holeIndices for earcut
+      const holeIndices: number[] = [];
+      while (primitivePolygonIndices[primitiveIndex] < endIdx) {
+        if (primitivePolygonIndices[primitiveIndex] > startIdx) {
+          holeIndices.push(primitivePolygonIndices[primitiveIndex] - startIdx);
+        }
+        primitiveIndex++;
+      }
+      const triangleIndices = earcut(
+        slicedFlatCoords,
+        holeIndices.length > 0 ? holeIndices : undefined,
+        nDim
+      );
+      if (triangleIndices.length === 0) {
+        throw Error('earcut failed e.g. invalid polygon');
```

Comment:
Using ogr2ogr to convert e.g. GeoJSON to Arrow doesn't fix invalid polygons and align the geometry types. Therefore, in many cases, there are invalid polygons and mismatched polygon types (MultiPolygons in a Simple Polygon arrow file), which cause earcut hanging or returning empty indices.

Reply:
> MultiPolygons in a Simple Polygon arrow file

in geoarrow that can't really exist; either the whole column is a multipolygon or polygon type. So you should be able to see from the metadata whether the column is a multipolygon or polygon type